RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTHcalVarProducerFromRecHit.cc

```diff
@@ -0,0 +1,233 @@
+    if (absEtaLowEdges_.size() != effectiveAreas_.size()) {
+      throw cms::Exception("IncompatibleVects") << "absEtaLowEdges and effectiveAreas should be of the same size. \n";
+    }
+
+    if (absEtaLowEdges_.at(0) != 0.0) {
+      throw cms::Exception("IncompleteCoverage") << "absEtaLowEdges should start from 0. \n";
+    }
+
+    for (unsigned int aIt = 0; aIt < absEtaLowEdges_.size() - 1; aIt++) {
+      if (!(absEtaLowEdges_.at(aIt) < absEtaLowEdges_.at(aIt + 1))) {
+        throw cms::Exception("ImproperBinning") << "absEtaLowEdges entries should be in increasing order. \n";
+      }
+    }
+  }
+}
+
+void EgammaHLTHcalVarProducerFromRecHit::fillDescriptions(edm::ConfigurationDescriptions &descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("recoEcalCandidateProducer", edm::InputTag("hltRecoEcalCandidate"));
+  desc.add<edm::InputTag>("rhoProducer", edm::InputTag("fixedGridRhoFastjetAllCalo"));
+  desc.add<edm::InputTag>("hbheRecHitsTag", edm::InputTag("hltHbhereco"));
+  desc.add<bool>("doRhoCorrection", false);
+  desc.add<double>("rhoMax", 999999.);
+  desc.add<double>(("rhoScale"), 1.0);
+  //eThresHB/HE are from RecoParticleFlow/PFClusterProducer/python/particleFlowRecHitHBHE_cfi.py
+  desc.add<std::vector<double> >("eThresHB", {0.1, 0.2, 0.3, 0.3});
+  desc.add<std::vector<double> >("etThresHB", {0, 0, 0, 0});
+  desc.add<std::vector<double> >("eThresHE", {0.1, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2});
+  desc.add<std::vector<double> >("etThresHE", {0, 0, 0, 0, 0, 0, 0});
+  desc.add<double>("innerCone", 0);
+  desc.add<double>("outerCone", 0.14);
+  desc.add<int>("depth", 0);
+  desc.add<int>("maxSeverityHB", 9);
+  desc.add<int>("maxSeverityHE", 9);
+  desc.add<bool>("doEtSum", false);
+  desc.add<bool>("useSingleTower", false);
+  desc.add<std::vector<double> >("effectiveAreas", {0.079, 0.25});  // 2016 post-ichep sinEle default
+  desc.add<std::vector<double> >("absEtaLowEdges", {0.0, 1.479});   // Barrel, Endcap
+  descriptions.add("hltEgammaHLTHcalVarProducerFromRecHit", desc);
+}
+
+void EgammaHLTHcalVarProducerFromRecHit::produce(edm::StreamID,
+                                                 edm::Event &iEvent,
+                                                 const edm::EventSetup &iSetup) const {
+  auto recoEcalCandHandle = iEvent.getHandle(recoEcalCandidateProducer_);
+
+  double rho = 0.0;
+
+  if (doRhoCorrection_) {
+    rho = iEvent.get(rhoProducer_);
+    if (rho > rhoMax_) {
+      rho = rhoMax_;
+    }
+    rho = rho * rhoScale_;
+  }
+
+  reco::RecoEcalCandidateIsolationMap isoMap(recoEcalCandHandle);
+
+  for (unsigned int iRecoEcalCand = 0; iRecoEcalCand < recoEcalCandHandle->size(); iRecoEcalCand++) {
+    reco::RecoEcalCandidateRef recoEcalCandRef(recoEcalCandHandle, iRecoEcalCand);
+
+    float isol = 0;
+    EgammaHcalIsolation::InclusionRule external;
+    EgammaHcalIsolation::InclusionRule internal;
+
+    if (!useSingleTower_) {  //useSingleTower_=False means H/E is cone-based
```

Comment:
Something less error prone in case of future updates could be:

```suggestion
    if (useSingleTower_) {
      if (!doEtSum_) {  //this is single tower based H/E
        external = EgammaHcalIsolation::InclusionRule::isBehindClusterSeed;
        internal = EgammaHcalIsolation::InclusionRule::withinConeAroundCluster;
      } else {  //this is cone-based HCAL isolation with single tower based footprint removal
        external = EgammaHcalIsolation::InclusionRule::withinConeAroundCluster;
        internal = EgammaHcalIsolation::InclusionRule::isBehindClusterSeed;
    } else {  //useSingleTower_=False means H/E is cone-based
      external = EgammaHcalIsolation::InclusionRule::withinConeAroundCluster;
      internal = EgammaHcalIsolation::InclusionRule::withinConeAroundCluster;
    }    
```

Reply:
thank you @perrotta , I've now addressed your comments